### DISCUSSION TITLE
Expose `bufferedAmount`

### DISCUSSION THREAD

**Discussion by Raynos** [sockjs/sockjs-client#559] [Ideas]
[`bufferedAmount`](http://www.whatwg.org/specs/web-apps/current-work/multipage/network.html#dom-websocket-bufferedamount) exposes how much of the data is being buffered right now.

If you want to implement some kind of backpressure mechanism with say high and low watermarks you would need to check `bufferedAmount` > high mark and apply back pressure then emit some kind of `"drain"` event when `bufferedAmount` < low mark.

To do this sockJS needs to expose this information.

**Comment by majek** [comment]
This is a very good point. I guess you need that both on the client and server side?

**Comment by Raynos** [comment]
@majek I don't need it on the server. Any compliant node websocket client should already have it.

**Comment by arlolra** [comment]
do _any_ node websocket clients expose `bufferedAmount`?

**Comment by Raynos** [comment]
They should. Websockets do. Not sure of any ws abstraction that exposes it.

**Comment by arlolra** [comment]
@Raynos `should !== do`. For instance, [ws](https://github.com/einaros/ws) doesn't. [faye](https://github.com/faye/faye-websocket-node) does, though it's a constant, maybe indicating nothing is buffered? Anyways, I'm looking for a compliant library, if you come across any.

**Comment by Raynos** [comment]
@arlolra it's trivial to get access to it by reading `socket._writableState.length` for TCP sockets

**Comment by dmnsgn** [comment]
Is there an equivalent concept to `bufferedAmount` for xhr?

**Comment by brycekahle** [comment]
Not that I'm aware of.
